mgr/dashboard: smb join-auth and usersgroups resources listing

Pegonzal opened the pull request:
Fixes: https://tracker.ceph.com/issues/69361


[screen-capture (56).webm](https://github.com/user-attachments/assets/95e85e17-7d63-417e-8f2e-64fa06d7ca66)



<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

dnyanee1997 commented:
@Pegonzal , thanks for the PR. 
I like this idea of having Join Auth and User Group functionality as seperate.
I have tested the PR it's working as expected.

Pegonzal commented:
jenkins test make check

nmunet commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,26 @@
+<ul class="nav nav-tabs">
```
As we are moving towards carbon, would you mind changing this component to use carbon tabs?

nmunet reviewed (commented):
Hi @Pegonzal,
Changes looks good to me, I have put a minor suggestion that you can look into!
Thanks

nizamial09 commented on src/pybind/mgr/dashboard/controllers/smb.py:
```diff
@@ -87,6 +87,41 @@
     }, "Configuration for the CephFS share")
 }
 
+JOIN_AUTH_SCHEMA = {
+    "resource_type": (str, "ceph.smb.join.auth"),
+    "auth_id": (str, "Unique identifier for the join auth resource"),
+    "intent": (str, "Desired state of the resource, e.g., 'present' or 'removed'"),
+    "auth": ({
+        "username": (str, "Username for authentication"),
+        "password": (str, "Password for authentication")
+    }, "Authentication credentials"),
+    "linked_to_cluster": (str, "Optional string containing a cluster ID. \
+    If set, the resource is linked to the cluster and will be automatically removed \
+    when the cluster is removed")
+}
+
+LIST_JOIN_AUTH_SCHEMA = [JOIN_AUTH_SCHEMA]
+
+USERSGROUPS_SCHEMA = {
+    "resource_type": (str, "ceph.smb.usersgroups"),
+    "users_groups_id": (str, "A short string identifying the usersgroups resource"),
+    "intent": (str, "Desired state of the resource, e.g., 'present' or 'removed'"),
+    "values": ({
+        "users": ([{
+            "name": (str, "The user name"),
+            "password": (str, "The password for the user")
+        }], "List of user objects, each containing a name and password"),
+        "groups": ([{
+            "name": (str, "The name of the group")
+        }], "List of group objects, each containing a name")
+    }, "Required object containing users and groups information"),
+    "linked_to_cluster": (str, "Optional string containing a cluster ID. \
+    If set, the resource is linked to the cluster and will be automatically removed \
+    when the cluster is removed")
+}
+
+LIST_USERSGROUPS_SCHEMA = [USERSGROUPS_SCHEMA]
+
```
schema keeps on growing ;)

Pegonzal replied:
Yeah... I realized I can at least generate other needed schemas from just a main one for each resource, so I'll be doing that.
I have it in another commit https://github.com/ceph/ceph/pull/61315/files#diff-293f6bd157fd219cf75ef48e0d571752e8bee7354bfd38921f2733c6e1f55426R96

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-list/smb-join-auth-list.component.ts:
```diff
@@ -0,0 +1,64 @@
+  ngOnInit() {
+    this.columns = [
+      {
+        name: $localize`ID`,
+        prop: 'auth_id',
+        flexGrow: 2
```
i found it a bit weird that it only has 1 column. But I remember you mentioning that you are looking to see more things to expose here? maybe separate PR?

Pegonzal replied:
It looks a bit wierd indeed... But these resources, [join-auth](https://docs.ceph.com/en/latest/mgr/smb/#join-auth-resource) and [usersgroups](https://docs.ceph.com/en/latest/mgr/smb/#users-and-groups-resource) don't have much else on them other than ID and usernames/passwords 

For now I've only listed the ID since I know it's something we want for sure, but will try to get the feedback next time I demo this to whether to show usernames or not (or anything else) and how to show them, maybe on a dropdown with a details page?

Also @nizamial09 let me know your thoughs if you have any suggestions

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,39 @@
+    i18n>
+    Users Groups
+  </a>
+</ng-template>
```
nit: no newline

Pegonzal replied:
Fixed it

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.spec.ts:
```diff
@@ -0,0 +1,22 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
tests for heading of tabs , we can add

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb.model.ts:
```diff
@@ -26,3 +26,38 @@ export interface AuthMode {
+export interface UsersGroups {
+  resource_type: string;
+  users_groups_id: string;
+  intent: 'present' | 'removed';
```
I see this is being used multiple places, we can create a type or enum .

Pegonzal replied:
Added type `Intent`

afreen23 reviewed (changes requested):
I think we don't need to use CLI based names as it is. Its quite verbose and as a user I need to look at documentation.

We can certainly improve here:

* Adding helper texts for each
* Instead of `join-auth` we can call Active Directory as you are also using in bread crumbs
* Similarly usersgroups resource , we can just say "Users" and list their username, password and groups associated in column.

But I also want to understand if they can be connected to a SMB cluster somehow ? Like a particular user using a cluster only ? if true then we should add those informations too.

I would like to discuss and understand more offline.
Thanks @Pegonzal

Pegonzal commented:
> I think we don't need to use CLI based names as it is. Its quite verbose and as a user I need to look at documentation. UI should be intuitive.

Hey @afreen23, I agree with you, I just didn't know what naming to go for instead so I just used the same as the smb module.

> * Instead of `join-auth` we can call Active Directory/Domain/or whatever widely known term as you are also using in bread crumbs (Active Directory so might be that)
> * Similarly usersgroups resource , we can just say "Users" and list their username, password and groups associated in column.

As I mentioned in the PR description, I am open to suggestions, and was trying get some feedback on this. 
Maybe we could go with: `Active Directory` for `join-auth` and `Standalone`  for `usersgroups`
And with a header/helper-text/tooltip that something along the lines of: 
`Active Directory` -> `Manage active-directory domain user access for your cluster`
`Standalone` -> `Manage standalone user groups for your cluster`

> 
> But I also want to understand if they can be connected to a SMB cluster somehow ? Like a particular user using a cluster only ? if true then we should add those information too.
> 

Good point, this resources (join-auth and usersgroups) will be used to access the cluster. Hence clusters need to know the ID of one of this resources (this is specified on cluster creation) and they'll store this id. So we could get the clusters information and list which cluster is associated the resource with, is that something we want to do?
FYI: A join-auth or usersgroup resource could be associated to multiple clusters

> I would like to discuss and understand more offline. Thanks @Pegonzal

Sure @afreen23

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-list/smb-join-auth-list.component.spec.ts:
```diff
@@ -0,0 +1,27 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
Maybe, but not really sure what to test from here, any suggestions?
For now the service request for listing this, is tested

Pegonzal commented:
Hi @afreen23, please take another look, I've addressed your comments and updated the video with the new terminology

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-domain-setting-modal/smb-domain-setting-modal.component.html:
```diff
@@ -41,6 +41,31 @@
+                i18n>-- Select an access resource --
+              </option>
+              <option
+                *ngFor="let jAId of joinAuthsIds"
```
```suggestion
                *ngFor="let jaID of joinAuthsIds"
```

afreen23 replied:
```suggestion
                *ngFor="let jAId of joinAuthsIdList"
```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-list/smb-join-auth-list.component.ts:
```diff
@@ -0,0 +1,62 @@
+  tableActions: CdTableAction[];
+  context: CdTableFetchDataContext;
+
+  // joinAuth$: Observable<SMBJoinAuth[]>;
+  // subject$ = new BehaviorSubject<SMBJoinAuth[]>([]);
```
```suggestion

```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-list/smb-join-auth-list.component.ts:
```diff
@@ -0,0 +1,62 @@
+        flexGrow: 2
+      },
+      {
+        name: $localize`Username`,
+        prop: 'auth.username',
+        flexGrow: 2
+      },
+      {
+        name: $localize`Password`,
+        prop: 'auth.password',
+        flexGrow: 2
+      },
```
we  can add copy to clipboard here for ease of access to user

Pegonzal replied:
removed passwords as for now due to feedback gotten last demo, might add them later hidden by default with an option to see them

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,29 @@
+    ></cd-smb-join-auth-list>
+  </cds-tab>
+  <cds-tab
+    heading="Standalone"
```
for each of these tabs we should add one line help text here, thanks/ Doc team can help here

Pegonzal replied:
What do you exactly mean by line help text, the `cd-helper` component? How would that work with the tab?

I was trying to set a tooltip for the tab, but could not make it work, it was not showing the content, in the video can be seen very small that pointer of the tooltip is showing when hovering over the tab but nothing else.

[tooltipnotshowing.webm](https://github.com/user-attachments/assets/88d28ebc-ce0e-4baa-9ef3-c1c0f80703cf)


But anyways, since I've updated it to display a whole header for the page for each of the tabs, I think we don't really need the tooltips

nizamial09 replied:
again taking an example from nvmeof component, maybe this is what @afreen23 was talking about? But I don't have a strong preference on whether the Heading and text is inside the tab content or outside it. I prefer consistency more or less.

![image](https://github.com/user-attachments/assets/d6bf4748-6fac-495d-83eb-09c6acd6c8cb)

afreen23 replied:
Yes we can do that way. @Pegonzal lets do same way of if there is a carbon way do in that manner. Thanks

afreen23 replied:
We should achieve consistency.
Also @Pegonzal you may check with UX on how to put. What we did with nvme was not something cross verified.
So outside/inside lets clarify with UX team please. Thanks for all gr8 work

Pegonzal replied:
I talked with UX, I am waiting on them to get their feedback back to me. As per their estimations, this Thursday I should have updates regarding the feedback.

For now, I'm following the same pattern as in NFS, except that I'm udpating the text displayed depending on the active tab

afreen23 replied:
The problem here is consistency. I really dont like the fact we are disregarding consistency across product. As far as current implementation is concerned this is not like nvmeof. Moving ahead I would prefer to see consistent carbonized components.

For here, if we choose tab headings inside tab then let it be for all why go away with that ?
And if we choosing here outside tab then have it outside for all tabs and fix other places similarly.

Thanks

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-details/smb-usersgroups-details.component.html:
```diff
@@ -0,0 +1,4 @@
+<cd-table-key-value
+    [data]="usernamePassword"
+    [customColNames]="customColNames"
+></cd-table-key-value>
```
nit: no newline

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,62 @@
+  *ngIf="selectedTab == Tabs.clusters"
+  i18n
+>
+  <h1>Clusters</h1>
+  <h5>Logical management units that may map to one or more managed Samba service</h5>
```
localization should be on the closest tags instead of parent for better string extraction. Same for below and other places.
```suggestion
  <h1 i18n>Clusters</h1>
  <h5 i18n>Logical management units that may map to one or more managed Samba service</h5>
```

Pegonzal replied:
Oh okay, I didn't know. I have updated it

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.spec.ts:
```diff
@@ -0,0 +1,38 @@
+  });
+
+  // If the cache is actvie it will try to fetch for every tab at once
+  // the smb module does not like that when trying to request the smb db
```
This comment seems confusing, can we improve ?

Pegonzal replied:
Changed it from:
```
// If the cache is actvie it will try to fetch for every tab at once
 // the smb module does not like that when trying to request the smb db
```

To now:
```
 // If the tabs cacheActive is set to true data for all tabs will be fetched at once,
 // smb mgr module might hit recursive error when doing multiple request to the db
  ```
  
  Hopefully that is more clear, and about this issue, I'll let John now.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-list/smb-join-auth-list.component.ts:
```diff
@@ -0,0 +1,69 @@
+        flexGrow: 2
+      },
+      {
+        name: $localize`Linked to`,
```
As it is, it might not be clear where its linked 

```suggestion
        name: $localize`Linked to Cluster`,
```

Pegonzal replied:
Sure, updated it.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,54 @@
+<div *ngIf="selectedTab == Tabs.clusters">
+  <h1 i18n>Clusters</h1>
```
bit skeptical on the heading as h1. It looks really big. Bigger than what we used to see. Probably we should define the title with a class (hopefully one that is there in the carbon) and then reuse that wherever we want this title to appear.

I saw that we have this format in [nvmeof component](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems/nvmeof-subsystems.component.html#L18)

![image](https://github.com/user-attachments/assets/e32ff582-8104-44d5-bb1a-2f3c1a0da102)
this looks a bit more nicer and it takes use of the `cd-help-text` to show the info text which also looks fitting in the page. WDYT?

Pegonzal replied:
Ah, I wasn't aware we already had something similar in place. Thanks for the heads up, the example you prived looks good, so I'll follow that then.

I've updated it to follow the same pattern.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-tabs/smb-tabs.component.html:
```diff
@@ -0,0 +1,54 @@
+<div
+  *ngIf="selectedTab == Tabs.joinAuths"
+>
+  <h1 i18n>Active Directory access resources</h1>
```
```suggestion
  <h1 i18n>Active directory access resources</h1>
```

Pegonzal replied:
Updated it

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-list/smb-usersgroups-list.component.html:
```diff
@@ -0,0 +1,25 @@
+  #groupsNamesTpl
+  let-row="data.row"
+>
+  <span *ngFor="let group of row.values.groups; let isLast=last">
+    {{ group.name }}{{ isLast ? '' : ', ' }}
+  </span>
```
for a better visualization, the badges or the carbon alternate, tags?
https://carbondesignsystem.com/components/tag/usage/

move to carbon tags is still something in progress though: https://github.com/ceph/ceph/pull/60361

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-list/smb-usersgroups-list.component.html:
```diff
@@ -0,0 +1,25 @@
+</ng-container>
+
+<ng-template
+  #groupsNamesTpl
```
nit
```suggestion
  #groupNamesTpl
```

Pegonzal replied:
Actually, I've been following the naming of the backend variables, and this is for the `SMB Usersgroups` both users and groups as plural since this resource can have multiple users and multiple groups. 
So leaving it is as it is.

Ref: https://docs.ceph.com/en/latest/mgr/smb/#users-and-groups-resource

nizamial09 replied:
alright, thanks

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-list/smb-usersgroups-list.component.ts:
```diff
@@ -0,0 +1,83 @@
+})
+export class SmbUsersgroupsListComponent extends ListWithDetails implements OnInit {
+  @ViewChild('groupsNamesTpl', { static: true })
+  groupsNamesTpl: TemplateRef<any>;
```
```suggestion
  groupNamesTpl: TemplateRef<any>;
```

Pegonzal replied:
Same reasoning as before, following naming convention for this resource as: `usersgroups`

Ref: https://docs.ceph.com/en/latest/mgr/smb/#users-and-groups-resource

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-list/smb-usersgroups-list.component.ts:
```diff
@@ -0,0 +1,83 @@
+  tableActions: CdTableAction[];
+  context: CdTableFetchDataContext;
+
+  usersGroups$: Observable<SMBUsersGroups[]>;
```
more nit!
```suggestion
  userGroups$: Observable<SMBUsersGroups[]>;
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-usersgroups-list/smb-usersgroups-list.component.ts:
```diff
@@ -0,0 +1,83 @@
+        flexGrow: 2
+      },
+      {
+        name: $localize`Linked to`,
```
```suggestion
        name: $localize`Linked to cluster`,
```

Pegonzal replied:
Updated it.

nizamial09 replied:
maybe you missed to update it.

Pegonzal replied:
Yes, I did missed it... I changed it only for join-auth-list. Need to change it here as well. If there are no more requested changes for this PR, I would prefer to go with it and I'll make this change over this other PR #61315 (which is a follow up to this one). Mainly to avoid re runnig the jenkin tests

nizamial09 reviewed (approved):
Great job @Pegonzal. Functionally looks perfect. Some nits though

nizamial09 commented:
oh i forgot, the content looks misaligned from what it shows here: https://carbondesignsystem.com/components/tabs/usage/#contained-tabs

as i understand it, if we want to have it like that we might need to override the cds--tab--content style to make sure its padding is 0.

Pegonzal commented:
> oh i forgot, the content looks misaligned from what it shows here: https://carbondesignsystem.com/components/tabs/usage/#contained-tabs
> 
> as i understand it, if we want to have it like that we might need to override the cds--tab--content style to make sure its padding is 0.

Thanks @nizamial09, I think we could apply that globally for tabs that have a table as a child (as their content). I'll bring that up for discussion next Dashboard standup, anyways that could be tackled in any of the other open PRs.

I've addressed your comments from your last review, I think it's looking better now, thanks!

[screen-capture (56).webm](https://github.com/user-attachments/assets/fa19162c-7dd3-48b1-a86b-a38f61184229)

Pegonzal commented:
Hi @afreen23, I addressed your requested changes. Can we merge this once tests are passing?

Pegonzal commented:
jenkins test make check arm64

Pegonzal commented:
jenkins test make check arm64

Pegonzal commented:
jenkins test make check arm64

afreen23 reviewed (approved):
Approving the PR based on carbon guidelines https://carbondesignsystem.com/components/tabs/usage/#contained-tabs we can have outside so lets do other places, @Pegonzal please create one tracker under "Usability" thanks.
